File: nhitomi-idx/nhitomi/Discord/ListJumpTrigger.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Discord;

namespace nhitomi.Discord
{
    public enum ListJumpTriggerDestination
    {
        Start,
        End,
        Input
    }

    public interface IListJumpTriggerTarget : IListTriggerTarget
    {
        int Start => 0;
        int End { get; }
    }

    public class ListJumpTrigger : ReactionTrigger
    {
        readonly IListJumpTriggerTarget _target;
        readonly ListJumpTriggerDestination _destination;

        public override IEmote Emote => _destination switch
        {
            ListJumpTriggerDestination.Start => new Emoji("\u23EA"),
            ListJumpTriggerDestination.End   => new Emoji("\u23E9"),
            ListJumpTriggerDestination.Input => new Emoji("\ud83d\udcd1"),

            _ => null
        };

        public ListJumpTrigger(IListJumpTriggerTarget target, ListJumpTriggerDestination destination)
        {
            _target      = target;
            _destination = destination;
        }

        protected override async Task<ReactionTriggerResult> RunAsync(CancellationToken cancellationToken = default)
        {
            var l = Context.Locale.Sections["reactions.list"];

            switch (_destination)
            {
                case ListJumpTriggerDestination.Start:
                    if (await _target.SetPositionAsync(_target.Start, cancellationToken))
                        return ReactionTriggerResult.Handled | ReactionTriggerResult.StateUpdated;

                    break;

                case ListJumpTriggerDestination.End:
                    if (await _target.SetPositionAsync(_target.End, cancellationToken))
                        return ReactionTriggerResult.Handled | ReactionTriggerResult.StateUpdated;

                    break;

                case ListJumpTriggerDestination.Input:
                    var message = await Message.ListenAsync(l["jump"], cancellationToken);

                    if (int.TryParse(message.Content, out var page) && await _target.SetPositionAsync(Math.Clamp(page - 1, _target.Start, _target.End), cancellationToken))
                        return ReactionTriggerResult.Handled | ReactionTriggerResult.StateUpdated;

                    break;
            }

            return ReactionTriggerResult.Ignored;
        }
    }
}

File: nhitomi-discord/src/Triggers/ListJumpTrigger.ts
import { ReactionTrigger } from "../Interactive/trigger";
import { ListTriggerTarget } from "./ListTrigger";
import { InteractiveMessage } from "../Interactive/message";

function getEmoji(destination: "start" | "end" | "input") {
  switch (destination) {
    case "start":
      return "\u23EA";

    case "end":
      return "\u23E9";

    case "input":
      return "\ud83d\udcd1";
  }
}

export type ListJumpTriggerTarget = ListTriggerTarget & {
  length: number;
};

export class ListJumpTrigger extends ReactionTrigger {
  constructor(
    readonly interactive: InteractiveMessage & ListJumpTriggerTarget,
    readonly destination: "start" | "end" | "input"
  ) {
    super(interactive, getEmoji(destination));
  }

  protected async run() {
    switch (this.destination) {
      case "start":
        this.interactive.position = 0;
        break;

      case "end":
        this.interactive.position = this.interactive.length - 1;
        break;

      case "input": {
        const result = parseInt(
          (await this.interactive.waitInput(this.context.locale.get("reaction.list.jump"))) || ""
        );

        if (isNaN(result)) {
          return false;
        }

        this.interactive.position = result - 1;
        break;
      }
    }

    return true;
  }
}
